src/mono/wasm/runtime/binding_support.js

```diff
@@ -2115,12 +2375,208 @@ var BindingSupportLib = {
+
+			return task_ptr;
+		}
+		catch (ex) {
+			return BINDING.wrap_exception(is_exception, ex);
+		}
+		finally {
+			uri_root.release();
+			sub_root.release();
+			on_close_root.release();
+		}
+	},
+	mono_wasm_web_socket_send: function (webSocket_js_handle, message_ptr, length, message_type, end_of_message, thenable_js_handle, is_exception) {
+		var message_root = MONO.mono_wasm_new_root(message_ptr);
+		try {
+			const ws = BINDING.mono_wasm_get_jsobj_from_js_handle(webSocket_js_handle)
+			if (!ws)
+				throw new Error("ERR17: Invalid JS object handle " + webSocket_js_handle);
 
+			const buffer = BINDING._mono_wasm_web_socket_send_buffering(ws, message_ptr, length, message_type, end_of_message);
+
+			if (!end_of_message) {
+				return null; // we are done buffering synchronously, no promise
+			}
+			return BINDING._mono_wasm_web_socket_send_and_wait(ws, buffer, thenable_js_handle);
+		}
+		catch (ex) {
+			return BINDING.wrap_exception(is_exception, ex);
+		}
+		finally {
+			message_root.release();
+		}
+	},
+	mono_wasm_web_socket_receive: function (webSocket_js_handle, buffer_ptr, length, thenable_js_handle, is_exception) {
+		var buffer_root = MONO.mono_wasm_new_root(buffer_ptr);
+		try {
+			const ws = BINDING.mono_wasm_get_jsobj_from_js_handle(webSocket_js_handle)
+			if (!ws)
+				throw new Error("ERR18: Invalid JS object handle " + webSocket_js_handle);
+			var event_queue = ws[BINDING.wasm_ws_pending_receive_event_queue];
+			var promise_queue = ws[BINDING.wasm_ws_pending_receive_promise_queue];
+
+			if (event_queue.getLength()) {
+				// finish synchronously
+				const res = BINDING._mono_wasm_web_socket_receive_buffering(event_queue, buffer_root, length);
+
+				setValue(thenable_js_handle, 0, "i32");
+				return BINDING._task_from_result(res);
+			}
+
+			const { promise, promise_control } = BINDING._create_cancelable_promise();
+			promise_control.buffer_root = buffer_root;
+			promise_control.buffer_length = length;
+			promise_queue.enqueue(promise_control);
+
+			var { task_ptr, then_js_handle } = BINDING._wrap_js_thenable_as_task(promise);
+			setValue(thenable_js_handle, then_js_handle, "i32");
+			return task_ptr;
+		}
+		catch (ex) {
+			return BINDING.wrap_exception(is_exception, ex);
+		}
+	},
+	mono_wasm_web_socket_close: function (webSocket_js_handle, code, reason, wait_for_close_received, thenable_js_handle, is_exception) {
+		var reason_root = MONO.mono_wasm_new_root(reason);
+		try {
+			const ws = BINDING.mono_wasm_get_jsobj_from_js_handle(webSocket_js_handle)
+			if (!ws)
+				throw new Error("ERR19: Invalid JS object handle " + webSocket_js_handle);
+
+			var js_reason = BINDING.conv_string(reason_root.value);
+
+			if (wait_for_close_received) {
+				const { promise, promise_control } = BINDING._create_cancelable_promise();
+				ws[BINDING.wasm_ws_pending_close_promise] = promise_control;
+
+				ws.addEventListener('close', (ev) => {
+					promise_control.resolve(ev.code);
```

Comment:
in one of the other cases we pass through ev.reason, we're passing through ev.code here. we should be consistent, maybe pass through `closed reason=${ev.reason} code=${ev.code}`

Reply:
I removed this way of passing that code+reason as we already have `js_on_close` callback registered too.
I unified to the same "close" listener, to make sure it's called in the right order.